mne/cov.py

```diff
@@ -1585,33 +1617,36 @@ def compute_whitener(noise_cov, info, picks=None, rank=None,
         The whitening matrix.
     ch_names : list
         The channel names.
+    rank : int
+        Rank reduction of the whitener. Returned only if return_rank is True.
     """
     if picks is None:
         picks = pick_types(info, meg=True, eeg=True, ref_meg=False,
                            exclude='bads')
 
-    ch_names = [info['chs'][k]['ch_name'] for k in picks]
+    ch_names = [info['ch_names'][k] for k in picks]
 
-    # This function does not modify inplace
-    noise_cov = prepare_noise_cov(noise_cov, info, ch_names,
-                                  rank=rank, scalings=scalings)
-    n_chan = len(ch_names)
+    # XXX this relies on pick_channels, which does not respect order,
+    # so this could create problems if users have reordered their data
+    noise_cov = pick_channels_cov(noise_cov, include=ch_names, exclude=[])
+    if len(noise_cov['data']) != len(ch_names):
+        missing = list(set(ch_names) - set(noise_cov['names']))
+        raise RuntimeError('Not all channels present in noise covariance:\n%s'
+                           % missing)
 
-    W = np.zeros((n_chan, n_chan), dtype=np.float)
-    #
-    #   Omit the zeroes due to projection
-    #
-    eig = noise_cov['eig']
-    nzero = (eig > 0)
-    W[nzero, nzero] = 1.0 / np.sqrt(eig[nzero])
-    #
-    #   Rows of eigvec are the eigenvectors
-    #
-    W = np.dot(W, noise_cov['eigvec'])
+    if diag:
+        noise_cov = noise_cov.copy()
+        noise_cov['data'] = np.diag(np.diag(noise_cov['data']))
+
+    scalings = _handle_default('scalings_cov_rank', scalings)
```

Comment:
I think this line is not needed any longer. It anyways reappears further down the road in the cov rank estimation function (which we will probably abolish in the future).

Reply:
I will keep it for now so you can test I can safely remove it. you need to redo the evaluation you did when you added this.